src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md

```diff
@@ -0,0 +1,549 @@
+---
+title: React 如何区分 Class 和 Function？
+date: '2018-12-02'
+spoiler: 我们谈论关于类（classes）、new、instanceof、原型链（prototype chains）、和 API 设计。
+
+---
+
+让我们来看一下这个以函数形式定义的 `Greeting` 组件：
+
+```jsx
+function Greeting() {
+  return <p>Hello</p>;
+}
+```
+
+React 也支持将他定义成一个类：
+
+```jsx
+class Greeting extends React.Component {
+  render() {
+    return <p>Hello</p>;
+  }
+}
+```
+
+（直到 [最近](https://reactjs.org/docs/hooks-intro.html)，这是使用 state 特性的唯一方式）
+
+当你要渲染一个 `<Greeting />` 组件时，你并不需要关心它是如何定义的：
+
+```jsx
+// 是类还是函数 —— 无所谓
+<Greeting />
+```
+
+但 *React 本身* 在意其中的差别！
+
+如果 `Greeting` 是一个函数，React 需要调用它。
+
+```jsx
+// 你的代码
+function Greeting() {
+  return <p>Hello</p>;
+}
+
+// React 内部
+const result = Greeting(props); // <p>Hello</p>
+```
+
+但如果 `Greeting` 是一个类，React 需要先用 `new` 操作符将其实例化，*然后* 调用刚才生成实例的 `render` 方法：
+
+```jsx
+// 你的代码
+class Greeting extends React.Component {
+  render() {
+    return <p>Hello</p>;
+  }
+}
+
+// React 内部
+const instance = new Greeting(props); // Greeting {}
+const result = instance.render(); // <p>Hello</p>
+```
+
+无论哪种情况 React 的目标都是去获取渲染后的节点（在这个案例中，`<p>Hello</p>`）。但具体的步骤取决于 `Greeting` 是如何定义的。
+
+**所以 React 是怎么知道某样东西是 class 还是 function 的呢？**
+
+就像我 [上一篇博客](https://overreacted.io/why-do-we-write-super-props/) 中提到的，**你并不 *需要* 知道这个才能高效使用 React。** 我多年来都不知道这个。请不要把这变成一道面试题。事实上，这篇博客更多的是关于 JavaScript 而不是 React。
+
+这篇博客是写给那些对 React 具体是 **如何** 工作的表示好奇的读者的。你是那样的人吗？那我们一起深入探讨一下吧。
+
+**这将是一段漫长的旅程，系好安全带。这篇文章并没有多少关于 React 本身的信息，但我们会涉及到 `new`、`this`、`class`、箭头函数、`prototype`、`__proto__`、`instanceof` 等方面，以及这些东西是如何在 JavaScript 中一起工作的。幸运的是，你并不需要在使用 React 时一直想着这些，除非你正在实现 React...**
+
+（如果你真的很想知道答案，直接翻到最下面。）
+
+* * *
+
+首先，我们需要理解为什么把函数和类分开处理很重要。注意看我们是怎么使用 `new` 操作符来调用一个类的：
+
+```jsx
+// 如果 Greeting 是一个函数
+const result = Greeting(props); // <p>Hello</p>
+
+// 如果 Greeting 是一个类
+const instance = new Greeting(props); // Greeting {}
+const result = instance.render(); // <p>Hello</p>
+```
+
+我们来简单看一下 `new` 在 JavaScript 是干什么的。
+
+* * *
+
+在过去，JavaScript 还没有类。但是，你可以使用普通函数来模拟。**具体来讲，只要在函数调用前加上 `new` 操作符，你就可以把 *任何* 函数当做一个类的构造函数来用：**
+
+```jsx
+// 只是一个函数
+function Person(name) {
+  this.name = name;
+}
+
+var fred = new Person('Fred'); // ✅ Person {name: 'Fred'}
+var george = Person('George'); // 🔴 没用的
+```
+
+现在你依然可以这样写！在 DevTools 里试试吧。
+
+如果你调用 `Person('Fred')` 时 **没有** 加 `new`，其中的 `this` 会指向某个全局且无用的东西（比如，`window` 或者 `undefined`），因此我们的代码会崩溃，或者做一些像设置 `window.name` 之类的傻事。
+
+通过在调用前增加 `new`，我们说：“嘿 JavaScript，我知道 `Person` 只是个函数，但让我们假装它是个构造函数吧。**创建一个 `{}` 对象并把 `Person` 中的 `this` 指向那个对象，以便我可以通过类似 `this.name` 的形式去设置一些东西，然后把这个对象返回给我。**”
```

Comment:
```suggestion
通过在调用前增加 `new`，我们说：「嘿 JavaScript，我知道 `Person` 只是个函数，但让我们假装它是个构造函数吧。**创建一个 `{}` 对象并把 `Person` 中的 `this` 指向那个对象，以便我可以通过类似 `this.name` 的形式去设置一些东西，然后把这个对象返回给我。**」
```
I suggest using full-width corner brackets `「` and `」`. See below for reference. 
https://github.com/mzlogin/chinese-copywriting-guidelines#%E5%85%A8%E8%A7%92%E5%92%8C%E5%8D%8A%E8%A7%92